File: benchmarks/Java/src/richards/HandlerTaskDataRecord.java
/*
 * This benchmark is derived from Mario Wolczko's Java and Smalltalk version of
 * Richards.
 * 
 * It is modified based on the SOM version and to use Java 8 features.
 * License details:
 *   http://web.archive.org/web/20050825101121/http://www.sunlabs.com/people/mario/java_benchmarking/index.html
 */
package richards;


final class HandlerTaskDataRecord extends RBObject {
  private Packet workIn;
  private Packet deviceIn;

  HandlerTaskDataRecord() {
    workIn = deviceIn = NO_WORK;
  }

  public Packet deviceIn() { return deviceIn; }
  public void deviceIn(final Packet aPacket) { deviceIn = aPacket; }

  public void deviceInAdd(final Packet packet) {
    deviceIn = append(packet, deviceIn);
  }

  public Packet workIn() { return workIn; }
  public void workIn(final Packet aWorkQueue) { workIn = aWorkQueue; }

  public void workInAdd(final Packet packet) {
    workIn = append(packet, workIn);
  }
}

File: benchmarks/Rust/src/richards/handler_task_data_record.rs
use crate::richards::packet::{Packet, PacketBox};
use crate::richards::rb_object::RBObject;

pub struct HandlerTaskDataRecord {
    work_in: Option<PacketBox>,
    device_in: Option<PacketBox>,
}

impl Default for HandlerTaskDataRecord {
    fn default() -> Self {
        Self {
            work_in: Packet::NO_WORK,
            device_in: Packet::NO_WORK,
        }
    }
}

impl RBObject for HandlerTaskDataRecord {}

impl HandlerTaskDataRecord {
    pub fn take_device_in(&mut self) -> Option<PacketBox> {
        self.device_in.take()
    }

    pub fn set_device_in(&mut self, device_in: Option<PacketBox>) {
        self.device_in = device_in;
    }

    pub fn device_in_add(&mut self, packet: PacketBox) {
        self.device_in = Some(Self::append(packet, self.device_in.take()));
    }

    pub fn take_work_in(&mut self) -> Option<PacketBox> {
        self.work_in.take()
    }

    pub fn set_work_in(&mut self, work_in: Option<PacketBox>) {
        self.work_in = work_in;
    }

    pub fn work_in_add(&mut self, packet: PacketBox) {
        self.work_in = Some(Self::append(packet, self.work_in.take()));
    }
}
